File: packages/collection-view/ios/Support/AppearanceResolver.swift
import UIKit

/**
 * Resolves the light/dark appearance pair into UIKit values.
 *
 * The important design decision is that **colours come back as dynamic `UIColor`s**, built with
 * `UIColor(dynamicProvider:)`. UIKit then re-resolves them itself whenever the trait collection
 * changes, which means a light/dark switch needs no reconfiguration of any cell, no trait
 * observation for the common case, and — crucially — no round trip to JavaScript. The naive
 * alternative, resolving to a concrete colour at apply time, produces a list that only restyles
 * on the *next* unrelated render, which is the kind of bug that gets reported as "dark mode is
 * flaky".
 *
 * A field left unset falls through to the platform's own colour, and those are already dynamic,
 * so partial theming stays correct in both modes for free.
 */
struct AppearanceResolver {
  private let light: Appearance?
  private let dark: Appearance?

  init(light: Appearance?, dark: Appearance?) {
    self.light = light
    self.dark = dark
  }

  /**
   * A colour that follows the interface style, or `fallback` when neither side sets it.
   *
   * `dark` falls back to `light` field by field, never the other way round: setting only
   * `appearance` should give that look in both modes, which is the least surprising behaviour
   * and means adding a dark override is always additive.
   */
  func color(_ keyPath: KeyPath<Appearance, String?>, fallback: UIColor) -> UIColor {
    optionalColor(keyPath, fallback: fallback) ?? fallback
  }

  /**
   * `nil` when neither side themes this field, which is not the same as "the default colour".
   *
   * `UIBackgroundConfiguration` is the reason this distinction has to exist: its
   * `backgroundColor` is `nil` by default and that nil is *meaningful* — it lets UIKit supply
   * both the resting colour and the highlight behaviour. Assigning a concrete colour, even the
   * identical one, replaces that behaviour and the row stops greying out when pressed. So an
   * unthemed field must be left alone rather than resolved to its own default.
   */
  func optionalColor(
    _ keyPath: KeyPath<Appearance, String?>,
    fallback: UIColor? = nil
  ) -> UIColor? {
    let lightHex = light?[keyPath: keyPath]
    let darkHex = dark?[keyPath: keyPath] ?? lightHex

    guard lightHex != nil || darkHex != nil else { return nil }

    return UIColor { traits in
      let hex = traits.userInterfaceStyle == .dark ? darkHex : lightHex
      // Resolving the fallback against the same traits keeps a partially themed list coherent:
      // an unparseable value behaves exactly as if the field had been left out.
      return UIColor(rnguiHex: hex)
        ?? fallback?.resolvedColor(with: traits)
        ?? .clear
    }
  }

  /**
   * Non-colour values, resolved against a specific trait collection.
   *
   * These cannot be dynamic — there is no `UIFont(dynamicProvider:)` and no dynamic `CGFloat` —
   * so a caller who genuinely sets a different font or spacing per mode needs the visible cells
   * reconfigured when the style flips. That is the only reason this component observes trait
   * changes at all; see `interfaceStyleDidChange` in the host.
   *
   * Takes a closure rather than a `KeyPath<Appearance, T?>`. The key-path form looks tidier and
   * type-checks, but unifying `KeyPath<Appearance, FontSpec?>` against `KeyPath<Appearance, T?>`
   * resolved to something that returned nil at runtime for a value that was demonstrably present
   * in the decoded struct. A closure leaves nothing to infer.
   */
  func value<T>(
    _ pick: (Appearance) -> T?,
    for traits: UITraitCollection
  ) -> T? {
    if traits.userInterfaceStyle == .dark, let dark, let value = pick(dark) {
      return value
    }
    return light.flatMap(pick)
  }

  /// True when either side sets a value that cannot be expressed as a dynamic colour, and the
  /// list therefore has to be reconfigured on a style change rather than left to UIKit.
  var hasStyleDependentNonColorValues: Bool {
    guard let dark else { return false }
    return dark.font != nil
      || dark.headerFont != nil
      || dark.footerFont != nil
      || dark.sectionSpacing != nil
  }
}

File: packages/collection-view/android/src/main/java/com/rngui/collectionview/AppearanceResolver.kt
package com.rngui.collectionview

import android.content.Context
import android.content.res.Configuration
import androidx.annotation.AttrRes
import androidx.annotation.ColorInt
import androidx.appcompat.view.ContextThemeWrapper
import com.google.android.material.R as MaterialR
import com.google.android.material.color.MaterialColors
import com.rngui.collectionview.generated.Appearance
import com.rngui.collectionview.generated.FontSpec

/** What the `colorScheme` prop pins the list to. */
enum class ColorScheme {
  system,
  light,
  dark;

  companion object {
    fun from(raw: String?): ColorScheme =
      when (raw) {
        "light" -> light
        "dark" -> dark
        else -> system
      }
  }
}

/**
 * Resolves the light/dark appearance pair against the mode in force, over Material 3 tokens.
 *
 * **The defaults are M3 colour roles, not iOS colours transliterated.** An unthemed list draws with
 * `surface`, `surfaceContainer`, `onSurface`, `onSurfaceVariant`, `outlineVariant` and `primary` —
 * which means it follows the consuming app's Material theme, including dynamic colour on Android 12+,
 * rather than looking like an iOS list that happens to run here. Anything the caller sets in
 * `appearance` still wins; the tokens are what "unset" means.
 *
 * **This is where Android and iOS genuinely differ, rather than differing in spelling.** The Swift
 * resolver hands back dynamic `UIColor`s and lets UIKit re-resolve them on a trait change, with no
 * cell reconfigured. Android has no dynamic colour — a resolved colour is an `Int`, and an `Int`
 * does not know what mode produced it — so the same guarantee is bought by rebinding instead.
 *
 * `dark` falls back to `light` per field, never the other way round: setting only `appearance`
 * gives that look in both modes, so adding a dark override is always additive.
 */
class AppearanceResolver(
  /**
   * Already wrapped for the mode in force — see [themedContext]. Every token below resolves
   * against *this* context's theme, which is what makes `colorScheme` work at all.
   */
  val context: Context,
  val isDark: Boolean,
  private val light: Appearance?,
  private val dark: Appearance?,
) {
  private fun <T> pick(get: (Appearance) -> T?): T? {
    if (isDark) dark?.let(get)?.let { return it }
    return light?.let(get)
  }

  /** A themed colour, or the M3 role behind it when neither side sets one. */
  @ColorInt
  fun color(get: (Appearance) -> String?, role: String, @ColorInt fallback: Int): Int =
    optionalColor(get) ?: token(role, fallback)

  /**
   * `null` when neither side themes this field, which is not the same as "the default colour".
   *
   * A row whose background is left alone keeps its `RippleDrawable` and therefore its press
   * feedback; assigning a concrete colour — even the identical one — replaces the drawable and the
   * row stops responding to touch. Unthemed has to mean untouched.
   */
  @ColorInt
  fun optionalColor(get: (Appearance) -> String?): Int? = parseRnguiHex(pick(get))

  /**
   * An M3 colour role off the themed context, looked up by *name*.
   *
   * **By name rather than through `com.google.android.material.R.attr`, and that is not a
   * stylistic choice.** A library's R fields are non-final and rewritten when the app is
   * assembled; with non-transitive R classes the app's regenerated `material.R$attr` need not
   * carry every field this library compiled against, and the result is a runtime
   * `NoSuchFieldError: No field colorPrimary` on the first row it draws. Declaring our own `<attr>`
   * references instead fails the resource merge outright — the names are already defined.
   *
   * Resolving by name asks the merged resource table the same question, at runtime, where the
   * answer is always right. Cached per name, so it costs one lookup per role per process.
   */
  @ColorInt
  fun token(role: String, @ColorInt fallback: Int): Int {
    val attr = attrId(context, role)
    if (attr == 0) return fallback
    return MaterialColors.getColor(context, attr, fallback)
  }

  /** Non-colour values. No dynamic form exists for these on either platform. */
  fun font(get: (Appearance) -> FontSpec?): FontSpec? = pick(get)

  fun dimension(get: (Appearance) -> Double?): Double? = pick(get)

  companion object {
    private val attrIds = HashMap<String, Int>()

    @AttrRes
    private fun attrId(context: Context, name: String): Int =
      attrIds.getOrPut(name) {
        context.resources.getIdentifier(name, "attr", context.packageName)
      }

    // The M3 colour roles this library draws with. Names rather than ids; see `token`.
    const val COLOR_PRIMARY = "colorPrimary"
    const val COLOR_ON_SURFACE = "colorOnSurface"
    const val COLOR_ON_SURFACE_VARIANT = "colorOnSurfaceVariant"
    const val COLOR_OUTLINE_VARIANT = "colorOutlineVariant"
    const val COLOR_SECONDARY_CONTAINER = "colorSecondaryContainer"
    const val COLOR_ON_SECONDARY_CONTAINER = "colorOnSecondaryContainer"
    const val COLOR_SURFACE = "colorSurface"
    const val COLOR_SURFACE_CONTAINER = "colorSurfaceContainer"

    fun isDark(configuration: Configuration, scheme: ColorScheme): Boolean =
      when (scheme) {
        ColorScheme.light -> false
        ColorScheme.dark -> true
        ColorScheme.system ->
          (configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK) ==
            Configuration.UI_MODE_NIGHT_YES
      }

    /**
     * A context carrying a Material 3 theme and the mode this list is resolving against.
     *
     * **Both halves are load-bearing, and neither is optional any more.**
     *
     * The *theme* is what `MaterialSwitch`, `MaterialCheckBox` and `MaterialRadioButton` require —
     * they read `colorPrimary`, `colorOnSurface` and a dozen shape attributes at inflation time and
     * throw without them — and it is where every M3 token this library draws with comes from.
     * Supplying it here rather than requiring one of the consuming app is the difference between a
     * library that works in any React Native project and one that comes with setup instructions.
     *
     * The *configuration override* is the Android answer to `overrideUserInterfaceStyle`. A theme
     * resolves `colorSurface` against its context's configuration, so overriding `uiMode` on the
     * wrapper makes every token below it resolve as though the device were in that mode — which is
     * exactly what `colorScheme` promises.
     *
     * `applyOverrideConfiguration` has to happen before anything reads a resource from the wrapper,
     * which is why the theme is set *after* it rather than in the constructor.
     */
    fun themedContext(base: Context, isDark: Boolean): Context {
      val configuration =
        Configuration(base.resources.configuration).apply {
          uiMode =
            (uiMode and Configuration.UI_MODE_NIGHT_MASK.inv()) or
              if (isDark) Configuration.UI_MODE_NIGHT_YES else Configuration.UI_MODE_NIGHT_NO
        }

      return ContextThemeWrapper(base, 0).apply {
        applyOverrideConfiguration(configuration)
        setTheme(MaterialR.style.Theme_Material3_DayNight_NoActionBar)
      }
    }
  }
}
